<a href="https://colab.research.google.com/github/1706712022/CasoPracticoElectiva2/blob/main/Notebooks/EJERCICIO2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

EJERCICIO 2 ELECTIVA 2 SE CARGAN LIBRERIAS

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

SE CARGAN LOS ARCHIVOS DESDE GIT

In [2]:
git_csv_urls = {
    "Ejercicio 1.csv": "https://raw.githubusercontent.com/1706712022/dataminig/refs/heads/main/cvs/Ejercicio%201.csv",
    "Ejercicio 2.csv": "https://raw.githubusercontent.com/1706712022/dataminig/refs/heads/main/cvs/Ejercicio%202.csv",
    "Ejercicio 3.csv": "https://raw.githubusercontent.com/1706712022/dataminig/refs/heads/main/cvs/Ejercicio%203.csv",
    "Ejercicio 4_Scoring_clean.csv": "https://raw.githubusercontent.com/1706712022/dataminig/refs/heads/main/cvs/Ejercicio%204_Scoring_clean.csv",
    "Ejercicio 4_Training_clean.csv": "https://github.com/your_repo/your_project/raw/main/Ejercicio 4_Training_clean.csv",
    "Ejercicio 5_clean.csv": "https://github.com/your_repo/your_project/raw/main/Ejercicio 5_clean.csv",
    "Ejercicio 6_Scoring.csv": "https://github.com/your_repo/your_project/raw/main/Ejercicio 6_Scoring.csv",
    "Ejercicio 6_Training.csv": "https://github.com/your_repo/your_project/raw/main/Ejercicio 6_Training.csv",
    "Ejercicio 7_Scoring.csv": "https://github.com/your_repo/your_project/raw/main/Ejercicio 7_Scoring.csv",
    "Ejercicio 7_Training.csv": "https://github.com/your_repo/your_project/raw/main/Ejercicio 7_Training.csv",
    "Ejercicio 8_Scoring.csv": "https://github.com/your_repo/your_project/raw/main/Ejercicio 8_Scoring.csv",
    "Ejercicio 8_Training.csv": "https://github.com/your_repo/your_project/raw/main/Ejercicio 8_Training.csv",
    "Ejercicio Importación.csv": "https://github.com/your_repo/your_project/raw/main/Ejercicio Importación.csv"
}

# You can access a URL like this:
# print(git_csv_urls["Ejercicio 1.csv"])


### Paso 2: Cargar datos
`df = pd.read_csv("Ejercicio 1.csv")`
`df.head()`

Se carga el dataset y se visualizan las primeras filas para validar su estructura.

In [4]:
df = pd.read_csv(git_csv_urls["Ejercicio 2.csv"])
display(df.head())

,Elapsed_Time,Time_in_Community,Gender,Working,Age,Family,Hobbies,Social_Club,Political,Professional,Religious,Support_Group
0,8.71,Short,M,No,53,1,0,0,0,0,0,0
1,5.24,Medium,F,No,31,0,0,0,0,0,1,1
2,4.22,Medium,M,No,42,1,1,0,0,1,0,0
3,4.81,Long,F,No,30,0,0,0,0,0,0,0
4,3.95,Long,M,Yes,29,0,0,0,1,1,0,1


### Paso 3: Inicial del dataset
`df.shape`

Cada registro corresponde a una respuesta individual registrada en la encuesta comunitaria. Cada columna representa una característica de la persona participante o su pertenencia a un tipo de organización.

In [7]:
df.shape

(3483, 12)

### Paso 4: Revision de columnas
`df.columns`

Las primeras variables describen información general de la persona participante, mientras que las últimas variables indican pertenencia a organizaciones. Estas últimas serán las más importantes para construir reglas de asociación.

In [8]:
df.columns

Index(['Elapsed_Time', 'Time_in_Community', 'Gender', 'Working', 'Age',
       'Family', 'Hobbies', 'Social_Club', 'Political', 'Professional',
       'Religious', 'Support_Group'],
      dtype='object')

### Paso 5: Vista Inicial de los Datos


In [10]:
df.head()

,Elapsed_Time,Time_in_Community,Gender,Working,Age,Family,Hobbies,Social_Club,Political,Professional,Religious,Support_Group
0,8.71,Short,M,No,53,1,0,0,0,0,0,0
1,5.24,Medium,F,No,31,0,0,0,0,0,1,1
2,4.22,Medium,M,No,42,1,1,0,0,1,0,0
3,4.81,Long,F,No,30,0,0,0,0,0,0,0
4,3.95,Long,M,Yes,29,0,0,0,1,1,0,1


### Paso 6: Revision de tipo de Datos

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3483 entries, 0 to 3482
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Elapsed_Time       3483 non-null   float64
 1   Time_in_Community  3483 non-null   object 
 2   Gender             3483 non-null   object 
 3   Working            3483 non-null   object 
 4   Age                3483 non-null   int64  
 5   Family             3483 non-null   int64  
 6   Hobbies            3483 non-null   int64  
 7   Social_Club        3483 non-null   int64  
 8   Political          3483 non-null   int64  
 9   Professional       3483 non-null   int64  
 10  Religious          3483 non-null   int64  
 11  Support_Group      3483 non-null   int64  
dtypes: float64(1), int64(8), object(3)
memory usage: 326.7+ KB


### Paso 7: Revision de valores Nulls

In [12]:
df.isnull().sum()

,0
Elapsed_Time,0
Time_in_Community,0
Gender,0
Working,0
Age,0
Family,0
Hobbies,0
Social_Club,0
Political,0
Professional,0


### Paso 8: Revision de Duplicados

In [13]:
df.duplicated().sum()

np.int64(5)

### Paso 9: Estadistica Descriptiva General

In [14]:
df.describe()

,Elapsed_Time,Age,Family,Hobbies,Social_Club,Political,Professional,Religious,Support_Group
count,3483.000000,3483.000000,3483.000000,3483.000000,3483.000000,3483.000000,3483.000000,3483.000000,3483.000000
mean,5.921786,36.730692,0.389894,0.300029,0.188056,0.093885,0.324433,0.418605,0.158771
std,2.293446,10.646524,0.487796,0.458336,0.390813,0.291710,0.468230,0.493401,0.365515
min,2.010000,17.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,3.875000,27.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
50%,5.930000,36.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
75%,7.840000,46.000000,1.000000,1.000000,0.000000,0.000000,1.000000,1.000000,0.000000
max,10.150000,57.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000


### Paso 10: Analisis de la variable Elapsep_time


In [15]:
df["Elapsed_Time"].describe()

,Elapsed_Time
count,3483.000000
mean,5.921786
std,2.293446
min,2.010000
25%,3.875000
50%,5.930000
75%,7.840000
max,10.150000


### Paso 11: Análisis de la variable Age

In [16]:
df["Age"].describe()

,Age
count,3483.000000
mean,36.730692
std,10.646524
min,17.000000
25%,27.000000
50%,36.000000
75%,46.000000
max,57.000000


#### DETECCION DE DATOS ATIPICOS

In [18]:
# Límite inferior = media - 2 * desviación estándar
# Límite superior = media + 2 * desviación estándar
variables_numericas = ["Elapsed_Time", "Age"]

for columna in variables_numericas:
    media = df[columna].mean()
    desviacion = df[columna].std()

    limite_inferior = media - 2 * desviacion
    limite_superior = media + 2 * desviacion

    outliers = df[
        (df[columna] < limite_inferior) |
        (df[columna] > limite_superior)
    ]

    print("Variable:", columna)
    print("Media:", round(media, 2))
    print("Desviación estándar:", round(desviacion, 2))
    print("Límite inferior:", round(limite_inferior, 2))
    print("Límite superior:", round(limite_superior, 2))
    print("Cantidad de valores atípicos:", len(outliers))
    print("-" * 40)

Variable: Elapsed_Time
Media: 5.92
Desviación estándar: 2.29
Límite inferior: 1.33
Límite superior: 10.51
Cantidad de valores atípicos: 0
----------------------------------------
Variable: Age
Media: 36.73
Desviación estándar: 10.65
Límite inferior: 15.44
Límite superior: 58.02
Cantidad de valores atípicos: 0
----------------------------------------


#### 7.14 Revisión de la variable Time_in_Community

In [19]:
df["Time_in_Community"].value_counts()

,count
Time_in_Community,
Long,1465
Medium,1304
Short,714


#### 7.15 Revisión de la variable Gender

In [20]:
df["Gender"].value_counts()

,count
Gender,
F,1790
M,1693


#### 7.16 Revisión de la variable Working

In [21]:
df["Working"].value_counts()

,count
Working,
Yes,1744
No,1739


#### 7.17 Selección de atributos para reglas de asociación


Para el modelo de reglas de asociación no se utilizarán todas las columnas. El objetivo del ejercicio es encontrar relaciones entre tipos de organizaciones. Por esa razón, se seleccionan únicamente las columnas que indican pertenencia a organizaciones.
•	¿Las personas que pertenecen a grupos religiosos también suelen pertenecer a grupos familiares?
•	¿Las personas que participan en grupos de hobbies también participan en organizaciones religiosas?
•	¿Qué tipos de organizaciones tienden a aparecer juntas?
Este paso sustituye al operador Select Attributes de RapidMiner.


In [22]:
columnas_asociacion = [
    "Family",
    "Hobbies",
    "Social_Club",
    "Political",
    "Professional",
    "Religious",
    "Support_Group"
]

#### 7.18 Creación del dataset de asociación

Se creó un nuevo DataFrame llamado df_asociacion. Este dataset contiene únicamente las variables relevantes para las reglas de asociación. Cada fila representa una respuesta individual registrada en la encuesta comunitaria y cada columna representa un tipo de organización.

In [23]:
df_asociacion = df[columnas_asociacion].copy()

df_asociacion.head()


,Family,Hobbies,Social_Club,Political,Professional,Religious,Support_Group
0,1,0,0,0,0,0,0
1,0,0,0,0,0,1,1
2,1,1,0,0,1,0,0
3,0,0,0,0,0,0,0
4,0,0,0,1,1,0,1


#### 7.19 Validación de valores binarios

Todas las columnas seleccionadas contienen únicamente valores binarios, lo que confirma que pueden convertirse correctamente a valores booleanos.

In [24]:
for columna in columnas_asociacion:
    print(columna, df_asociacion[columna].unique())


Family [1 0]
Hobbies [0 1]
Social_Club [0 1]
Political [0 1]
Professional [0 1]
Religious [0 1]
Support_Group [0 1]


#### 7.20 Conversión de 0 y 1 a valores booleanos

La conversión funciona así: 1 se convierte en True y 0 se convierte en False. Esto deja el dataset en un formato adecuado para aplicar reglas de asociación en Python. Este paso sustituye al operador Numerical to Binominal de RapidMiner.

In [25]:
df_asociacion_bool = df_asociacion.astype(bool)

df_asociacion_bool.head()


,Family,Hobbies,Social_Club,Political,Professional,Religious,Support_Group
0,True,False,False,False,False,False,False
1,False,False,False,False,False,True,True
2,True,True,False,False,True,False,False
3,False,False,False,False,False,False,False
4,False,False,False,True,True,False,True


#### 7.21 Verificación de tipos finales

Todas las variables quedaron convertidas al tipo bool. Esto confirma que el dataset está listo para el proceso de modelado mediante reglas de asociación.

In [26]:
df_asociacion_bool.dtypes

,0
Family,bool
Hobbies,bool
Social_Club,bool
Political,bool
Professional,bool
Religious,bool
Support_Group,bool


#### 7.22 Frecuencia de pertenencia a cada organización

La organización con mayor participación es Religious, con 1,458 registros. Luego aparecen Family, Professional y Hobbies. Las organizaciones con menor participación son Support_Group y Political.

In [27]:
frecuencias = df_asociacion.sum().sort_values(ascending=False)

frecuencias


,0
Religious,1458
Family,1358
Professional,1130
Hobbies,1045
Social_Club,655
Support_Group,553
Political,327


#### 7.23 Porcentaje de participación por organización

Los porcentajes muestran que las categorías con mayor presencia son Religious, Family, Professional y Hobbies. Este análisis ayuda a entender cuáles tipos de organizaciones tienen más peso dentro de la comunidad.

In [28]:
porcentajes = (df_asociacion.mean() * 100).sort_values(ascending=False)

porcentajes.round(2)


,0
Religious,41.86
Family,38.99
Professional,32.44
Hobbies,30.00
Social_Club,18.81
Support_Group,15.88
Political,9.39


#### 7.24 Revisión preliminar de combinaciones frecuentes

Las combinaciones con soporte mayor al 10% muestran relaciones preliminares interesantes. Las más relevantes son Hobbies + Religious, Family + Religious y Family + Hobbies. Esto indica que hay una frecuencia importante de respuestas donde estas organizaciones aparecen juntas.

In [30]:
from itertools import combinations

for grupo_a, grupo_b in combinations(columnas_asociacion, 2):
    soporte = (df_asociacion_bool[grupo_a] & df_asociacion_bool[grupo_b]).mean()

    if soporte > 0.10:
        print(grupo_a, "+", grupo_b, "=", round(soporte, 4))

Family + Hobbies = 0.1869
Family + Social_Club = 0.1289
Family + Professional = 0.1303
Family + Religious = 0.2245
Hobbies + Social_Club = 0.1235
Hobbies + Religious = 0.2389
Social_Club + Religious = 0.1473
Professional + Religious = 0.124


#### 7.25 Interpretación de las combinaciones preliminares


**Hobbies + Religious**

Esta combinación tiene el soporte más alto entre los pares revisados. Significa que aproximadamente el 23.89% de los registros pertenecen simultáneamente a grupos de hobbies y organizaciones religiosas. Esto puede indicar que personas activas en actividades recreativas también participan en comunidades religiosas.

**Family + Religious**

Esta combinación tiene un soporte de 22.45%. Significa que una proporción importante de personas participa tanto en organizaciones familiares como religiosas. Esto podría ser útil para coordinar proyectos sociales dirigidos a familias desde organizaciones religiosas.

**Family + Hobbies**

Esta combinación tiene un soporte de 18.69%. Significa que hay una relación preliminar entre personas que participan en organizaciones familiares y grupos de hobbies. Esto puede ser útil para diseñar actividades recreativas, familiares o comunitarias.


#### 7.26 Dataset final preparado

El dataset final contiene únicamente variables booleanas. Cada fila representa una respuesta individual registrada en la encuesta comunitaria. Cada columna representa un tipo de organización. El valor True indica que la persona pertenece a esa organización y el valor False indica que no pertenece.
Este dataset ya está preparado para aplicar algoritmos como Apriori, FP-Growth y Association Rules.


In [31]:
df_asociacion_bool.head()

,Family,Hobbies,Social_Club,Political,Professional,Religious,Support_Group
0,True,False,False,False,False,False,False
1,False,False,False,False,False,True,True
2,True,True,False,False,True,False,False
3,False,False,False,False,False,False,False
4,False,False,False,True,True,False,True


#### 7.27 Guardar el dataset preparado

Este comando guarda el dataset preparado en un nuevo archivo CSV. El archivo generado puede utilizarse posteriormente para el modelado de reglas de asociación.

In [32]:
df_asociacion_bool.to_csv("Ejercicio_2_preparado_asociacion.csv", index=False)